src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java

```diff
@@ -1,41 +1,483 @@
 	}
 
+	public static class SortEntry {
+		public String name;
+		// External interfaces only have access to ItemStacks, sorry fluids.
+		public Comparator<ItemStack> itemStackComparator;
+		// This is the internal comparator, fluids supported here.
+		public Comparator<IIngredientListElement> ingredientComparator;
+
+		public SortEntry(String name, Comparator<ItemStack> comparator) {
+			this.name = name;
+			this.itemStackComparator = comparator;
+		}
+
+		public SortEntry(String name) {
+			this.name = name;
+		}
+
+		public String getLocalisedName() {
+			return Translator.translateToLocal(name);
+		}
+
+		public String getTooltip() {
+			String tipname = name + ".tip";
+			String tip = Translator.translateToLocal(tipname);
+			return !tip.equals(tipname) ? tip : null;
+		}
+	}
+
 	@Override
 	public int compare(IIngredientListElement o1, IIngredientListElement o2) {
-		final String modName1 = o1.getModNameForSorting();
-		final String modName2 = o2.getModNameForSorting();
-
-		if (modName1.equals(modName2)) {
-			boolean isItemStack1 = (o1.getIngredient() instanceof ItemStack);
-			boolean isItemStack2 = (o2.getIngredient() instanceof ItemStack);
-			if (isItemStack1 && !isItemStack2) {
-				return -1;
-			} else if (!isItemStack1 && isItemStack2) {
-				return 1;
-			}
-
-			final int orderIndex1 = o1.getOrderIndex();
-			final int orderIndex2 = o2.getOrderIndex();
-			return Integer.compare(orderIndex1, orderIndex2);
-		} else if (modName1.equals(Constants.MINECRAFT_NAME)) {
-			return -1;
-		} else if (modName2.equals(Constants.MINECRAFT_NAME)) {
-			return 1;
-		} else {
-			return modName1.compareTo(modName2);
+		if (list.size() == 0) {
+			initConfig(); // emergency initializer, probably happened already.
+			// This should be the first time this gets hit after each config (re)load.
+			// This should allow enough time for other mods to load their options.
+			loadConfig(Config.getSortOrder());
+		}
+
+		for (SortEntry entry : list) {
+			int comparison = 0;
+			if (entry.ingredientComparator != null) {
+				comparison = entry.ingredientComparator.compare(o1, o2);
+			} else if (entry.itemStackComparator != null) {
+				ItemStack itemStack1 = getItemStack(o1);
+				ItemStack itemStack2 = getItemStack(o2);
+				if (itemStack1 == null && itemStack2 == null) {
+					comparison = 0;
+				} else if (itemStack1 != null && itemStack2 == null) {
+					comparison = -1;
+				} else if (itemStack1 == null && itemStack2 != null) {
+					comparison = 1;
+				} else {
+					comparison = entry.itemStackComparator.compare(itemStack1, itemStack2);
+				}
+			}
+			if (comparison != 0) {
+				return comparison;
+			}
+		}
+		return 0;
+	}
+
+	public static SortEntry find(String name) {
+		for (SortEntry entry : entries) {
+			if (entry.name.equals(name)) {
+				return entry;
+			}
+		}
+		return null;
+	}
+
+	public static int compareInt(int a, int b) {
+		return a == b ? 0 : a < b ? -1 : 1;
+	}
+	
+	public static int compareDouble(double a, double b) {
```

Comment:
you can use `java.lang.Double#compare` for this

Reply:
I cloned something I inherited.  I can do this.